rfc/rfc-0020-resources-update-after-success.md

```diff
@@ -0,0 +1,225 @@
+# Draft RFC 20 Read Resources Only When In Success State
+
+## Summary
+
+Cartographer is currently unable to connect an output of a resource to the inputs of said resource. Cartographer is
```

Comment:
In a conversation, @scothis brought up the following scenario:

Consider a supply chain with a sourceTemplate and a buildTemplate (where the sourceTemplate output is an input for the buildTemplate). Imagine the workload changes, with 2 updated values, one value that is an input for the sourceTemplate (perhaps a new source repository) and another that is an input for the imageTemplate (perhaps a new build env). While the sourceTemplate is reconciling, the imageTemplate will start reconciling on the new value from the workload. Later the sourceTemplate will finish reconciling and will pass its output to the buildTemplate. Can we disambiguate image output that is a result of the new workload val and _old_ source output from the image output that is a result of the new workload val and _new_ source?

Yes, we can! Indeed, this is an example of a general case, templates that have multiple inputs.

For the general case, Cartographer is responsible for going to a set of oracles and asking questions. Oracles give true answers. Cartographer wants to match answers to questions. By waiting until the oracle has answered all the questions posed, Cartographer can assure that the final answer is the answer to the final question.

What is necessary to make this assurance:
1. Cartographer must know the most recent question it has asked an oracle.
2. The oracle must be able to acknowledge that it has heard the question.
3. Having taken all questions and considered all observed states of the world, the oracle must have an "at rest" state where it shares an answer. This answer must be to the most recent question and must be correct based on the oracle's most recent observation of the world. If an oracle is in the process of answering a question and a new question is asked, the oracle is not "at rest" until it finishes answering the most recent question.
4. The oracle must be able to indicate when it is "at rest".
5. If the oracle can fail in answering the question, when "at rest" the oracle must indicate if it has succeeded or failed.

The case of Cartographer stamping an object from multiple inputs is simply asking a question with multiple parts. Ultimately we ask one question, we stamp one object. We know the input(s} that we used to stamp that object, we know the clause(s) of the question we ask. When an answer is provided by the oracle, when the object finishes reconciling, we can associate the answer with the question, the output with the input. There is nothing special about the input updating from the field of the workload.

---

Now, there are other concerns that an operator may have in the above scenario. Perhaps the two updates to the workload are meant to happen in lockstep. The intention is never to have an image produced with the new build env that doesn't also have the new source repository. That synchronization is out of scope of this RFC. Here we are only concerned with tracing; with giving users the ability to discover that a given output had certain inputs. This RFC would allow the operator to observe/discover the unsynchronized behavior, which is valuable.